{"pr_number": 2472, "pr_title": "HDFS-15689. allow/disallowSnapshot on EZ roots shouldn't fail due to trash provisioning/emptiness check", "pr_createdAt": "2020-11-18T10:32:04Z", "pr_url": "https://github.com/apache/hadoop/pull/2472", "timeline": [{"oid": "6f264681a45b10f4ad1f0fca1c23e5568872d35a", "url": "https://github.com/apache/hadoop/commit/6f264681a45b10f4ad1f0fca1c23e5568872d35a", "message": "HDFS-15689. allow/disallowSnapshot on EZ roots shouldn't fail due to trash provisioning/emptiness check\n\nChange-Id: Ic4bbdfad7aa4cb07082dc8b6ba1c50caa2769360", "committedDate": "2020-11-18T10:26:04Z", "type": "commit"}, {"oid": "2f03391493b7b98b0e8f335500a0160ac5bbc306", "url": "https://github.com/apache/hadoop/commit/2f03391493b7b98b0e8f335500a0160ac5bbc306", "message": "Add UT.\n\nChange-Id: Ib5ab219f384d4886cc5654216073a34ef1b14dfe", "committedDate": "2020-11-18T10:45:22Z", "type": "commit"}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUyNjAzNTQyOA==", "url": "https://github.com/apache/hadoop/pull/2472#discussion_r526035428", "bodyText": "I think there are 2 approaches instead of just ignoring the FileAlreadyExists exception:\n\nwe can just validate whether the existing Trash path is a directory and validate the permissions and not throw FileAlreadyExists exception itself in DistributedFileSystem.java#provisionSnapshotTrash\nIf the trash path already exists with right permissions, we can check if the path is an encryption zone as well and throw FileAlreadyExists exception only if its not an encryption zone. Similar change will be required for making an snapshottable dir an encryption zone.\n\nI am ok with either of the above approaches. I think just ignoring the exception here will not work in case. the existing path is not a directory or has right permissions.", "author": "bshashikant", "createdAt": "2020-11-18T12:05:17Z", "path": "hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/client/HdfsAdmin.java", "diffHunk": "@@ -171,7 +174,15 @@ public void clearQuotaByStorageType(Path src, StorageType type) throws IOExcepti\n   public void allowSnapshot(Path path) throws IOException {\n     dfs.allowSnapshot(path);\n     if (dfs.isSnapshotTrashRootEnabled()) {\n-      dfs.provisionSnapshotTrash(path, TRASH_PERMISSION);\n+      try {\n+        dfs.provisionSnapshotTrash(path, TRASH_PERMISSION);\n+      } catch (FileAlreadyExistsException ex) {\n+        // Don't throw on FileAlreadyExistsException since it is likely due to\n+        // admin allowing snapshot on an EZ root.\n+        LOG.warn(ex.getMessage());", "originalCommit": "2f03391493b7b98b0e8f335500a0160ac5bbc306", "replyToReviewId": null, "replies": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUyNjUzNjE0Mw==", "url": "https://github.com/apache/hadoop/pull/2472#discussion_r526536143", "bodyText": "Thanks @bshashikant for the comment.\nI was intending to not change the behavior of DFS#provisionSnapshotTrash but as I think again changing it should be fine since 3.4.0 is not released yet.\nI'm in favor of (1). It makes sense to reuse the trash if it is configured correctly. I will update the PR a bit later.", "author": "smengcl", "createdAt": "2020-11-19T01:40:47Z", "replyToReviewId": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUyNjAzNTQyOA=="}], "type": "inlineReview"}, {"oid": "6996910eb0c1b8f66a543ab79234517ea7a06f45", "url": "https://github.com/apache/hadoop/commit/6996910eb0c1b8f66a543ab79234517ea7a06f45", "message": "Don't throw if trash dir exists and permission matches.\n\nChange-Id: I03ec3bcc547be656b5d568853d477ee433c51978", "committedDate": "2020-11-23T11:29:59Z", "type": "commit"}, {"oid": "a07d917d574004adabe9cde9156aed40285626ef", "url": "https://github.com/apache/hadoop/commit/a07d917d574004adabe9cde9156aed40285626ef", "message": "Fix NPE in DFSClient#isEZRoot\n\nChange-Id: I11fb879dbf9a18bef286a1995506916dfd4fca60", "committedDate": "2020-11-23T19:09:37Z", "type": "commit"}]}